src/source/Common/EcsCredentialProvider.c

```diff
@@ -0,0 +1,290 @@
+/**
+ * Kinesis Video Producer ECS based Credential Provider
+ */
+#define LOG_CLASS "EcsCredentialProvider"
+#include "Include_i.h"
+
+STATUS getHostPort(PCHAR pUrl, PCHAR* ppStart, PCHAR* ppEnd)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PCHAR pStart = NULL, pEnd = NULL, pCurPtr;
+    UINT32 urlLen;
+    BOOL iterate = TRUE;
+
+    CHK(pUrl != NULL && ppStart != NULL && ppEnd != NULL, STATUS_NULL_ARG);
+
+    // We know for sure url is NULL terminated
+    urlLen = (UINT32) STRLEN(pUrl);
+
+    // Start from the schema delimiter
+    pStart = STRSTR(pUrl, PORT_DELIMITER_STRING);
+    CHK(pStart != NULL, STATUS_INVALID_ARG);
+
+    // Advance the pStart past the delimiter
+    pStart += STRLEN(PORT_DELIMITER_STRING);
+
+    // Ensure we are not past the string
+    CHK(pUrl + urlLen > pStart, STATUS_INVALID_ARG);
+
+    // Set the end first
+    pEnd = pUrl + urlLen;
+
+    // Find the delimiter which would indicate end of the host - either one of "/:?"
+    pCurPtr = pStart;
+    while (iterate && pCurPtr <= pEnd) {
+        switch (*pCurPtr) {
+            case '/':
+            case ':':
+            case '?':
+                iterate = FALSE;
+
+                // Set the new end value
+                pEnd = pCurPtr;
+            default:
+                pCurPtr++;
+        }
+    }
+
+CleanUp:
+
+    if (ppStart != NULL) {
+        *ppStart = pStart;
+    }
+
+    if (ppEnd != NULL) {
+        *ppEnd = pEnd;
+    }
+
+    return retStatus;
+}
+
+STATUS createEcsCredentialProviderWithTime(PCHAR ecsCredentialFullUri, PCHAR token, GetCurrentTimeFunc getCurrentTimeFn, UINT64 customData,
+                                           BlockingServiceCallFunc serviceCallFn, PAwsCredentialProvider* ppCredentialProvider)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PEcsCredentialProvider pEcsCredentialProvider = NULL;
+    PCHAR pStart, pEnd;
+    UINT32 len;
+    UINT32 port = 0;
+
+    CHK(ppCredentialProvider != NULL && ecsCredentialFullUri != NULL && token != NULL && serviceCallFn != NULL, STATUS_NULL_ARG);
+
+    pEcsCredentialProvider = (PEcsCredentialProvider) MEMCALLOC(1, SIZEOF(EcsCredentialProvider));
+    CHK(pEcsCredentialProvider != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pEcsCredentialProvider->credentialProvider.getCredentialsFn = getEcsCredentials;
+
+    // Store the time functionality and specify default if NULL
+    pEcsCredentialProvider->getCurrentTimeFn = (getCurrentTimeFn == NULL) ? commonDefaultGetCurrentTimeFunc : getCurrentTimeFn;
+    pEcsCredentialProvider->customData = customData;
+
+    CHK(STRNLEN(ecsCredentialFullUri, MAX_URI_CHAR_LEN + 1) <= MAX_URI_CHAR_LEN, MAX_URI_CHAR_LEN);
```

Comment:
Store the value of STRNLEN(ecsCredetialFullUri,..) and use it later

Reply:
Done